tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.james.mailbox.model.FetchGroup;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.store.StoreMailboxManager;
+import org.apache.james.task.Task;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.util.ReactorUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.linagora.tmail.james.jmap.settings.JmapSettings;
+import com.linagora.tmail.james.jmap.settings.JmapSettingsRepository;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.collection.JavaConverters;
+
+public class CleanupTrashService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CleanupTrashService.class);
+    public static final JmapSettingsValue WEEKLY_PERIOD = new JmapSettingsValue("weekly");
+    public static final JmapSettingsValue MONTHLY_PERIOD = new JmapSettingsValue("monthly");
+    public static final JmapSettingsKey TRASH_CLEANUP_ENABLED_SETTING = JmapSettings.JmapSettingsKey$.MODULE$.liftOrThrow("trash.cleanup.enabled");
+    public static final JmapSettingsKey TRASH_CLEANUP_PERIOD_SETTING = JmapSettings.JmapSettingsKey$.MODULE$.liftOrThrow("trash.cleanup.period");
+    private static final Map<JmapSettingsValue, Period> mapPeriodSettingToDuration = ImmutableMap.of(WEEKLY_PERIOD,  Period.ofWeeks(1),
+        MONTHLY_PERIOD, Period.ofMonths(1));
+
+    private final UsersRepository usersRepository;
+    private final JmapSettingsRepository jmapSettingsRepository;
+    private final StoreMailboxManager storeMailboxManager;
+    private final SystemMailboxesProvider systemMailboxesProvider;
+    private final Clock clock;
+
+    @Inject
+    public CleanupTrashService(UsersRepository usersRepository, JmapSettingsRepository jmapSettingsRepository, StoreMailboxManager storeMailboxManager,
+                               SystemMailboxesProvider systemMailboxesProvider, Clock clock) {
+        this.usersRepository = usersRepository;
+        this.jmapSettingsRepository = jmapSettingsRepository;
+        this.storeMailboxManager = storeMailboxManager;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash(RunningOptions runningOptions) {
+        return Flux.from(usersRepository.listReactive())
+            .transform(ReactorUtils.<Username, Task.Result>throttle()
+                .elements(runningOptions.getUsersPerSecond())
+                .per(Duration.ofSeconds(1))
+                .forOperation(this::cleanupTrashForSingleUser))
+            .reduce(Task.Result.COMPLETED, Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while accessing users from repository", e);
+                return Mono.just(Task.Result.PARTIAL);
```

Comment:
The errors can come from listing messages and cleaning up the messages phase, not only from accessing the UsersRepository.